From ddfd1900b26c66a062457d4fcc1a48bafd3eadf6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pavel=20B=C5=99ezina?= <pbrezina@redhat.com>
Date: Fri, 3 Feb 2017 13:04:23 +0100
Subject: [PATCH 25/79] cache_req: add ability to not use default domain suffix

This will be used in the next plugin "host by name" where
it is not desirable to use default domain suffix if set.

Reviewed-by: Jakub Hrozek <jhrozek@redhat.com>
---
 src/responder/common/cache_req/cache_req.c                        | 8 +++++++-
 src/responder/common/cache_req/cache_req_plugin.h                 | 5 +++++
 src/responder/common/cache_req/plugins/cache_req_enum_groups.c    | 1 +
 src/responder/common/cache_req/plugins/cache_req_enum_svc.c       | 1 +
 src/responder/common/cache_req/plugins/cache_req_enum_users.c     | 1 +
 .../common/cache_req/plugins/cache_req_group_by_filter.c          | 1 +
 src/responder/common/cache_req/plugins/cache_req_group_by_id.c    | 1 +
 src/responder/common/cache_req/plugins/cache_req_group_by_name.c  | 1 +
 .../common/cache_req/plugins/cache_req_initgroups_by_name.c       | 1 +
 .../common/cache_req/plugins/cache_req_initgroups_by_upn.c        | 1 +
 .../common/cache_req/plugins/cache_req_netgroup_by_name.c         | 1 +
 src/responder/common/cache_req/plugins/cache_req_object_by_id.c   | 1 +
 src/responder/common/cache_req/plugins/cache_req_object_by_name.c | 1 +
 src/responder/common/cache_req/plugins/cache_req_object_by_sid.c  | 1 +
 src/responder/common/cache_req/plugins/cache_req_svc_by_name.c    | 1 +
 src/responder/common/cache_req/plugins/cache_req_svc_by_port.c    | 1 +
 src/responder/common/cache_req/plugins/cache_req_user_by_cert.c   | 1 +
 src/responder/common/cache_req/plugins/cache_req_user_by_filter.c | 1 +
 src/responder/common/cache_req/plugins/cache_req_user_by_id.c     | 1 +
 src/responder/common/cache_req/plugins/cache_req_user_by_name.c   | 1 +
 src/responder/common/cache_req/plugins/cache_req_user_by_upn.c    | 1 +
 21 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/src/responder/common/cache_req/cache_req.c b/src/responder/common/cache_req/cache_req.c
index e5026e1a869064fe81cc04e3b2bbd8c4cefec304..aed8f1b225899a1c470407e259d2068ef62922b7 100644
--- a/src/responder/common/cache_req/cache_req.c
+++ b/src/responder/common/cache_req/cache_req.c
@@ -400,6 +400,7 @@ static errno_t cache_req_process_input(TALLOC_CTX *mem_ctx,
                                        const char *domain)
 {
     struct tevent_req *subreq;
+    const char *default_domain;
 
     if (cr->data->name.input == NULL) {
         /* Input was not name, there is no need to process it further. */
@@ -411,11 +412,16 @@ static errno_t cache_req_process_input(TALLOC_CTX *mem_ctx,
         return cache_req_set_name(cr, cr->data->name.input);
     }
 
+    default_domain = NULL;
+    if (!cr->plugin->ignore_default_domain) {
+        default_domain = cr->rctx->default_domain;
+    }
+
     /* Parse name since it may contain a domain name. */
     CACHE_REQ_DEBUG(SSSDBG_TRACE_FUNC, cr,
                     "Parsing input name [%s]\n", cr->data->name.input);
 
-    subreq = sss_parse_inp_send(mem_ctx, cr->rctx, cr->rctx->default_domain,
+    subreq = sss_parse_inp_send(mem_ctx, cr->rctx, default_domain,
                                 cr->data->name.input);
     if (subreq == NULL) {
         DEBUG(SSSDBG_CRIT_FAILURE, "Unable to create tevent request!\n");
diff --git a/src/responder/common/cache_req/cache_req_plugin.h b/src/responder/common/cache_req/cache_req_plugin.h
index e4d5eef91672a83e1ced47b394368a457acfbcb8..59ef8bad1697e094f729c53f33bda4f1d825cdff 100644
--- a/src/responder/common/cache_req/cache_req_plugin.h
+++ b/src/responder/common/cache_req/cache_req_plugin.h
@@ -157,6 +157,11 @@ struct cache_req_plugin {
     bool parse_name;
 
     /**
+     * True if default domain suffix should be ignored when parsing name.
+     */
+    bool ignore_default_domain;
+
+    /**
      * True if we always contact data provider.
      */
     bool bypass_cache;
diff --git a/src/responder/common/cache_req/plugins/cache_req_enum_groups.c b/src/responder/common/cache_req/plugins/cache_req_enum_groups.c
index de4bd968b18920cde0630dbd5142ce99d3b70a3e..2056dc2ccdadef98772402bde45aef8e043a0e76 100644
--- a/src/responder/common/cache_req/plugins/cache_req_enum_groups.c
+++ b/src/responder/common/cache_req/plugins/cache_req_enum_groups.c
@@ -64,6 +64,7 @@ const struct cache_req_plugin cache_req_enum_groups = {
     .dp_type = SSS_DP_GROUP,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = false,
+    .ignore_default_domain = false,
     .bypass_cache = true,
     .only_one_result = false,
     .search_all_domains = true,
diff --git a/src/responder/common/cache_req/plugins/cache_req_enum_svc.c b/src/responder/common/cache_req/plugins/cache_req_enum_svc.c
index c83564fdce8abc237a3a4dbe7a88b4bc6c2baaff..e850212977bb26dc13b900f6e5908865fffa59b0 100644
--- a/src/responder/common/cache_req/plugins/cache_req_enum_svc.c
+++ b/src/responder/common/cache_req/plugins/cache_req_enum_svc.c
@@ -65,6 +65,7 @@ const struct cache_req_plugin cache_req_enum_svc = {
     .dp_type = SSS_DP_SERVICES,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = false,
+    .ignore_default_domain = false,
     .bypass_cache = true,
     .only_one_result = false,
     .search_all_domains = true,
diff --git a/src/responder/common/cache_req/plugins/cache_req_enum_users.c b/src/responder/common/cache_req/plugins/cache_req_enum_users.c
index c4eeed7463cca6ecd17fe8042d62f4b72da46e68..2adeddb6b4bea044371f168f5d39aecc1f06cc45 100644
--- a/src/responder/common/cache_req/plugins/cache_req_enum_users.c
+++ b/src/responder/common/cache_req/plugins/cache_req_enum_users.c
@@ -64,6 +64,7 @@ const struct cache_req_plugin cache_req_enum_users = {
     .dp_type = SSS_DP_USER,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = false,
+    .ignore_default_domain = false,
     .bypass_cache = true,
     .only_one_result = false,
     .search_all_domains = true,
diff --git a/src/responder/common/cache_req/plugins/cache_req_group_by_filter.c b/src/responder/common/cache_req/plugins/cache_req_group_by_filter.c
index 1619cf7bdd6ad7ef7c1ea71ef0dd8f24611c1a6e..bc42eb7db0830ba31649c2cbb9525dfd1f7b1fae 100644
--- a/src/responder/common/cache_req/plugins/cache_req_group_by_filter.c
+++ b/src/responder/common/cache_req/plugins/cache_req_group_by_filter.c
@@ -119,6 +119,7 @@ const struct cache_req_plugin cache_req_group_by_filter = {
     .dp_type = SSS_DP_WILDCARD_GROUP,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = true,
+    .ignore_default_domain = false,
     .bypass_cache = true,
     .only_one_result = false,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_group_by_id.c b/src/responder/common/cache_req/plugins/cache_req_group_by_id.c
index 293994fa1e22a23b7ff19c50050e5c6c25274b5d..e48588087eafde68a4a85c546cf08e90eb6c7605 100644
--- a/src/responder/common/cache_req/plugins/cache_req_group_by_id.c
+++ b/src/responder/common/cache_req/plugins/cache_req_group_by_id.c
@@ -107,6 +107,7 @@ const struct cache_req_plugin cache_req_group_by_id = {
     .dp_type = SSS_DP_GROUP,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = false,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = true,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_group_by_name.c b/src/responder/common/cache_req/plugins/cache_req_group_by_name.c
index c88dbd4566297da98d306e20deb7f7c64c7991a4..962b38866a1408bbdff556e20df5a69b0d4bbba0 100644
--- a/src/responder/common/cache_req/plugins/cache_req_group_by_name.c
+++ b/src/responder/common/cache_req/plugins/cache_req_group_by_name.c
@@ -157,6 +157,7 @@ const struct cache_req_plugin cache_req_group_by_name = {
     .dp_type = SSS_DP_GROUP,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = true,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = true,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_initgroups_by_name.c b/src/responder/common/cache_req/plugins/cache_req_initgroups_by_name.c
index 9575ae70731875979f924dbf948222ed705fd923..d2f03cbea0780e4e0b88d56fcfbcf8903bcb3c85 100644
--- a/src/responder/common/cache_req/plugins/cache_req_initgroups_by_name.c
+++ b/src/responder/common/cache_req/plugins/cache_req_initgroups_by_name.c
@@ -172,6 +172,7 @@ const struct cache_req_plugin cache_req_initgroups_by_name = {
     .dp_type = SSS_DP_INITGROUPS,
     .attr_expiration = SYSDB_INITGR_EXPIRE,
     .parse_name = true,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = false,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_initgroups_by_upn.c b/src/responder/common/cache_req/plugins/cache_req_initgroups_by_upn.c
index 7a0b96b19f487e046c32235e02ec0fdbc7baa211..9b2d07d4afa98cbfca4a62f944b744f01897a0ee 100644
--- a/src/responder/common/cache_req/plugins/cache_req_initgroups_by_upn.c
+++ b/src/responder/common/cache_req/plugins/cache_req_initgroups_by_upn.c
@@ -108,6 +108,7 @@ const struct cache_req_plugin cache_req_initgroups_by_upn = {
     .dp_type = SSS_DP_INITGROUPS,
     .attr_expiration = SYSDB_INITGR_EXPIRE,
     .parse_name = false,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = false,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_netgroup_by_name.c b/src/responder/common/cache_req/plugins/cache_req_netgroup_by_name.c
index 15549adeff9e038387b21b6349b18683c14afe65..5b19edeb2952b83406ff20d001dd7d24449f69c9 100644
--- a/src/responder/common/cache_req/plugins/cache_req_netgroup_by_name.c
+++ b/src/responder/common/cache_req/plugins/cache_req_netgroup_by_name.c
@@ -116,6 +116,7 @@ const struct cache_req_plugin cache_req_netgroup_by_name = {
     .dp_type = SSS_DP_NETGR,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = true,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = true,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_object_by_id.c b/src/responder/common/cache_req/plugins/cache_req_object_by_id.c
index b8ad3b5e76cbf52fb61e22aa872e51e7f51bbf29..3f47807616054c644e27e4c240ad7c4b752a563e 100644
--- a/src/responder/common/cache_req/plugins/cache_req_object_by_id.c
+++ b/src/responder/common/cache_req/plugins/cache_req_object_by_id.c
@@ -99,6 +99,7 @@ const struct cache_req_plugin cache_req_object_by_id = {
     .dp_type = SSS_DP_USER_AND_GROUP,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = false,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = true,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_object_by_name.c b/src/responder/common/cache_req/plugins/cache_req_object_by_name.c
index 1ec906c7ad0c0f2d327667c697a96f2c2735d066..6829d0ec97c147aafda46b6eace25b97a28e626a 100644
--- a/src/responder/common/cache_req/plugins/cache_req_object_by_name.c
+++ b/src/responder/common/cache_req/plugins/cache_req_object_by_name.c
@@ -192,6 +192,7 @@ const struct cache_req_plugin cache_req_object_by_name = {
     .dp_type = SSS_DP_USER_AND_GROUP,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = true,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = true,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_object_by_sid.c b/src/responder/common/cache_req/plugins/cache_req_object_by_sid.c
index 35cb74f61fab0c72dda68c8f95e30be9127f938f..6a6eb8e72c52c069935ca4e612e60f602c7b91bd 100644
--- a/src/responder/common/cache_req/plugins/cache_req_object_by_sid.c
+++ b/src/responder/common/cache_req/plugins/cache_req_object_by_sid.c
@@ -109,6 +109,7 @@ const struct cache_req_plugin cache_req_object_by_sid = {
     .dp_type = SSS_DP_SECID,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = false,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = true,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_svc_by_name.c b/src/responder/common/cache_req/plugins/cache_req_svc_by_name.c
index 4de27571c199baeeec1064f6d9b626fef08212c7..9562354ed3a453e3aec7264bb32dbd5273fb0927 100644
--- a/src/responder/common/cache_req/plugins/cache_req_svc_by_name.c
+++ b/src/responder/common/cache_req/plugins/cache_req_svc_by_name.c
@@ -140,6 +140,7 @@ const struct cache_req_plugin cache_req_svc_by_name = {
     .dp_type = SSS_DP_SERVICES,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = true,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = false,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_svc_by_port.c b/src/responder/common/cache_req/plugins/cache_req_svc_by_port.c
index 1b17c71352678f7dfae830bea3ab3909fd62c564..55117492f6f8aa6a4e31c1e23862215255cdf660 100644
--- a/src/responder/common/cache_req/plugins/cache_req_svc_by_port.c
+++ b/src/responder/common/cache_req/plugins/cache_req_svc_by_port.c
@@ -113,6 +113,7 @@ const struct cache_req_plugin cache_req_svc_by_port = {
     .dp_type = SSS_DP_SERVICES,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = false,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = false,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_user_by_cert.c b/src/responder/common/cache_req/plugins/cache_req_user_by_cert.c
index 9a1bcc6aa1225c27362b11b9321994f65261d5cb..5203d3f94421715b711bcd1e01b7a42737b6fe41 100644
--- a/src/responder/common/cache_req/plugins/cache_req_user_by_cert.c
+++ b/src/responder/common/cache_req/plugins/cache_req_user_by_cert.c
@@ -83,6 +83,7 @@ const struct cache_req_plugin cache_req_user_by_cert = {
     .dp_type = SSS_DP_CERT,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = false,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = true,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_user_by_filter.c b/src/responder/common/cache_req/plugins/cache_req_user_by_filter.c
index ee9f60bf682629acf3b2ec3d16a3ed075084480d..4c328a5d900e37de0f3396a8c2f1c937360ce081 100644
--- a/src/responder/common/cache_req/plugins/cache_req_user_by_filter.c
+++ b/src/responder/common/cache_req/plugins/cache_req_user_by_filter.c
@@ -119,6 +119,7 @@ const struct cache_req_plugin cache_req_user_by_filter = {
     .dp_type = SSS_DP_WILDCARD_USER,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = true,
+    .ignore_default_domain = false,
     .bypass_cache = true,
     .only_one_result = false,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_user_by_id.c b/src/responder/common/cache_req/plugins/cache_req_user_by_id.c
index d710986d1102af4422d29a9943c903f23bea8b9e..d794d248b1e9b11cd41210b8180823e3a2565847 100644
--- a/src/responder/common/cache_req/plugins/cache_req_user_by_id.c
+++ b/src/responder/common/cache_req/plugins/cache_req_user_by_id.c
@@ -107,6 +107,7 @@ const struct cache_req_plugin cache_req_user_by_id = {
     .dp_type = SSS_DP_USER,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = false,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = true,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_user_by_name.c b/src/responder/common/cache_req/plugins/cache_req_user_by_name.c
index 46dd9434b34536b72b0966f53ab341c09542f16c..9ee7bef1cc904d25d156b3f64e039e47be58d1cc 100644
--- a/src/responder/common/cache_req/plugins/cache_req_user_by_name.c
+++ b/src/responder/common/cache_req/plugins/cache_req_user_by_name.c
@@ -157,6 +157,7 @@ const struct cache_req_plugin cache_req_user_by_name = {
     .dp_type = SSS_DP_USER,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = true,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = true,
     .search_all_domains = false,
diff --git a/src/responder/common/cache_req/plugins/cache_req_user_by_upn.c b/src/responder/common/cache_req/plugins/cache_req_user_by_upn.c
index 9d1e703d623cd830c2ab6db6e835c4bec49f57e5..4c6e6bcd056392abb729d416d406f28c28cdaa77 100644
--- a/src/responder/common/cache_req/plugins/cache_req_user_by_upn.c
+++ b/src/responder/common/cache_req/plugins/cache_req_user_by_upn.c
@@ -112,6 +112,7 @@ const struct cache_req_plugin cache_req_user_by_upn = {
     .dp_type = SSS_DP_USER,
     .attr_expiration = SYSDB_CACHE_EXPIRE,
     .parse_name = false,
+    .ignore_default_domain = false,
     .bypass_cache = false,
     .only_one_result = true,
     .search_all_domains = false,
-- 
2.9.3

